### PR TITLE
[ffigen] Retain listener blocks before invocation

### LINKED ISSUE #1570: ObjCBlock closures not registered with the objective_c `_blockClosureRegistry` in release build
During release build, usage of the ObjCBlock listener method that is generated by the ffigen tool causes an exception to be thrown. This is not present when using a debug build.

Crash Log
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Null check operator used on a null value
#0      getBlockClosure (package:objective_c/src/internal.dart:346)
#1      _ObjCBlock_ffiVoid_MLModel_NSError_closureTrampoline (package:path/to/binding/binding_gen.dart:31868)
#2      ObjCBlock_ffiVoid_MLModel_NSError.listener.<anonymous closure> (package:path/to/binding/binding_gen.dart)
#3      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594)
#4      _RootZone.bindUnaryCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1633)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184)
```

Code Snippet
```
  var completer = Completer<binding.MLModel?>();
  var handler = binding.ObjCBlock_ffiVoid_MLModel_NSError.listener(
      (binding.MLModel? model, objc.NSError? error) {
    completer.complete(model);
  });
  binding.MLModel.loadContentsOfURL_configuration_completionHandler_(
      compiledModel, config.config, handler);
  var model = await completer.future;
```

Versions
Dart: 3.5.3
Flutter: 3.24.3
objective_c: 2.0.0
ffi: 2.1.3
ffigen: 14:0

### LINKED ISSUE #1571: Random assertion failures `'_blockClosureRegistry.containsKey(id)': is not true`
In my `video_player` experiment, I not infrequently (especially when switching pages, so a player instance is being eliminated), get an assertion failure like this:

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: 'package:objective_c/src/internal.dart': Failed assertion: line 345 pos 10: '_blockClosureRegistry.containsKey(id)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:50:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      getBlockClosure (package:objective_c/src/internal.dart:345:10)
#3      _ObjCBlock_ffiVoid_closureTrampoline (package:video_player_avfoundation/src/ffi_bindings.dart:787:11)
#4      ObjCBlock_ffiVoid.listener.<anonymous closure> (package:video_player_avfoundation/src/ffi_bindings.dart)
#5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
#6      _RootZone.bindUnaryCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1633:26)
#7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

I may be doing something wrong, but I don't know what, or how to investigate/debug, since the exception doesn't give me any information. It seems like a lifetimer error, maybe? But I thought `.listener` would handle lifetime for me.

### PR DESCRIPTION
**Pull request by liamappelbe**
If an ObjC API holds the only reference to a Dart listener block, then the ObjC API invokes the block and immediately releases its reference to it, the invoke message and the destroy message will both be sent to the target isolate at about the same time. If the destroy message arrives first, it will delete the Dart function associated with the block before the invoke message arrives.

The fix is exactly the same as what we did for the listener's arguments. We retain a reference to the listener block in the ObjC-side trampoline before we invoke the listener, and the callback is responsible for releasing that reference. It used to be the case that we only generated these ObjC trampolines if one or more of the arguments needed to be retained, but now we have to generate the trampolines unconditionally, since the block itself needs to be retained. This actually simplified some of the logic.

As I mentioned on https://github.com/dart-lang/native/pull/1573, this bug is only reproducible in flutter, but the Dart GC function doesn't work on flutter. So for the test I'm using package:leak_tracker's `forceGC` function.

Also refactored the block bindings a bit to make them simpler and easier to read. Moved all the trampolines to be top level final variables and used the implicit lazy initialization of statics to lazy load them rather than explicitly using `??=`.

Fixes #1571. Fixes #1570.

### CODE REVIEW

File: pkgs/ffigen/test/native_objc_test/util.dart [hunk 1]
```diff
@@ -39,11 +40,19 @@ final _executeInternalCommand = () {
...
-  final gcNow = "gc-now".toNativeUtf8();
+  final gcNow = 'gc-now'.toNativeUtf8();
   _executeInternalCommand!(gcNow.cast(), nullptr);
   calloc.free(gcNow);
 }
 
+// Dart_ExecuteInternalCommand("gc-now") doesn't work on flutter, so we use
+// leak_tracker's forceGC function instead. It's less reliable, and to combat
+// that we need to wait for quite a long time, which breaks autorelease pools.
+Future<void> flutterDoGC() async {
```
Comment (HosseinYousefi) [inline pkgs/ffigen/test/native_objc_test/util.dart:51]: Interesting! What does leak tracker do to force GC and why isn't it reliable?

**Review by HosseinYousefi — APPROVED** [review]
LGTM!

File: pkgs/ffigen/test/native_objc_test/util.dart (hunk 1 above)
Comment (HosseinYousefi) [reply pkgs/ffigen/test/native_objc_test/util.dart:51]: Oh! It basically allocates a bunch of object!

Comment (liamappelbe) [reply pkgs/ffigen/test/native_objc_test/util.dart:51]: Yeah. The main issue is that it's async. In some of the tests I'm pushing an autorelease pool, doing a GC, then popping the pool. If the GC is async, then other stuff can run in between pushing and popping the pool. Flutter uses autorelease pools internally, so often they'll end up popping a pool that was pushed before mine, which breaks the stack push/pop pattern the pools are supposed to follow. This causes unexpected ref counts and occasional crashes.

**Pull request merged**